Added spec for agent health metrics (#708)
elastic/apm: specs/agents/metrics-health.md

 * When an item is removed from the queue, update `min_size` with the current queue size if it is less than `min_size`
 * When an item cannot be added to the queue (dropped), set `max_size` to `queue_capacity`

After the metrics are exported, `min_size` and `max_size` need to be reset tu the current queue size.

An atomic min/max can be implemented using the following algorithm:
```java
AtomicLong maxSize = new AtomicLong();

public void updateMaxSize(long currentSize) {
    //alternatively abort after like 10 iterations for constant latency
    while (true) { 
        long current = maxSize.get();
        if (current >= currentSize) {
            return;
        }
        boolean casSuccess = maxSize.compareAndSet(current, currentSize);
        if (casSuccess) {
            return;
        }
    }
}

```

### Event Request Metrics

Agents SHOULD expose the following metrics regarding Intake API networking:

 - `agent.events.requests.count`: `COUNTER` of the number of requests made or attempted to the Intake API of the APM server
 - `agent.events.requests.bytes`: `COUNTER` of the approximate number of bytes on the wire sent to the Intake API of the APM

> All these metrics MUST have the label `success`, which can have the values `true` or `false`. A request is counted with `success=true` iff the APM Server responded with `2xx`.

Reviewer: ```suggestion
The `agent.events.requests.*` metrics MUST have the label `success`, which can have the values `true` or `false`. A request is counted with `success=true` if the APM Server responded with `2xx`.
```

Author: `iff` stands for `if and only if` here, therefore it is not a typo.

The metric `agent.events.requests.bytes` does not need to represent the exact network usage.

> Instead the number of compressed bytes within the request body can be used as approximation.

Reviewer: ```suggestion
Instead, the number of compressed bytes within the request body can be used as approximation.
```

> All event request metrics can be disabled via the `disable_metrics` configuration.

Reviewer: ```suggestion
All event request metrics MUST be able to be disabled by the `disable_metrics` configuration option.
```

### Agent Overhead Metrics

> If possible with a reasonable amount of runtime and implementation overhead, agents SHOULD expose the following metrics for approximating their own resource usage:

Author: Agree, `MAY` is definitly the better wording here.

 - `agent.background.cpu.overhead.pct`: The fraction of the process cpu usage caused by the agent in a range [0-1] within the last reporting interval.
 - `agent.background.cpu.total.pct`: The approximate CPU usage caused by the agent. Derived by multiplying `agent.background.cpu.overhead.pct` with `system.process.cpu.total.norm.pct`.
 - `agent.background.memory.allocation.bytes`: `COUNTER` of the number of bytes allocated in the main memory by the agent.

>  - `agent.background.threads.count`: Number of currently running agent background threads at the time of reporting.

Reviewer: This came from some requirements docs. I think it just helps to give the confidence that there is no thread leak / excessive thread usage by the agents

In most cases, it will not be feasible to capture the overhead of every action the agent does due to the overhead of the measurement itself. E.g. for the Java agent it is not feasible to measure the overhead of the code added to user application methods via instrumentation. Instead, only the overhead of agent-owned threads is measured.

The `agent.background.cpu.overhead.pct` can be computed by measuring the increase of thread CPU time of a given task ([this method](https://docs.oracle.com/javase/7/docs/api/java/lang/management/ThreadMXBean.html#getCurrentThreadCpuTime()) for Java) and dividing it by the increase of the total process CPU time ([this method](https://docs.oracle.com/javase/7/docs/jre/api/management/extension/com/sun/management/OperatingSystemMXBean.html#getProcessCpuTime()) for Java) within the same period.

All agent overhead metrics are enabled/disabled via the `agent_background_overhead_metrics` configuration:

|                |   |
|----------------|---|
| Type           | `boolean` |
| Default        | `false` |
| Dynamic        | `false` |

> | Central config | `false` |

Author: > maybe another idea: have we considered changing the default of disable_metrics by adding agent.background.* to it for agents which implement overhead health metrics?

I considered it, but it reminded me of why the `experimental`-flag was separated from the `disable_instrumentations` configuration in the Java agent: If we add `agent.background.*` as a default to `disable_metrics` it is very easy for users to enable the overhead metrics by accident.
E.g. a user wants to disable some micrometer metrics, so he changes `disable_metrics=my.metrics.*`. Because he did not add the default `agent.background.*` as well, this unexpectedly enables the overhead metrics.

> I can imagine -especially for critical production apps- that users would love to turn on agent health metrics collection for some time (e.g. if they feel something is not ok with the agent), collect the agent health data, and then turn it off. All without restarting the agent (which typically means restarting their app).
>
> But this applies to both the overhead metrics (which are opt-in) and to the queue metrics (which are handled by disable_metrics if I understand it correctly). My point with this: I think if we make it dynamic (which would be useful), then all health metrics should be dynamic (not only one set of them).

I definitly agree with that it would be nice to have these configuration options dynamic, but thought aswell that it is an enhancement we could add later on. It would for example be great and simplify the configuration interface to just make `disable_metrics` dynamic in general.

> | Central config | `false` |

Author: Summary of the discussion in the java-agent-weekly:

We for now go with the additional configuration flag and later on remove / hide it when we have a better way of dealing with defaults of multi-value configs in general.

Many of these metrics are `COUNTER`s: a `COUNTER` is a monotonically increasing metric.
It is periodically reported by sending the difference of the counter-value now vs its value at the time of the previous report.
In order to save elasticsearch disk space, zero-increases are not sent.

Example:
| **Time** | **Counter** | **Reported Metric Value** |
|----------|-------------|---------------------------|
| 0s       | 7           | 7                         |
| 30s      | 10          | 3                         |
| 1m       | 10          | 0 (not sent)              |
| 1m30s    | 12          | 2                         |

### Event Count Metrics

Agents send their data to the Intake API of the APM-server in the form of a stream of events (e.g. transactions, spans, metricsets...).

The following metrics counting these events SHOULD be exposed:
 - `agent.events.total`: `COUNTER` of events either submitted to the internal reporting queue or dropped upon submit attempt
 - `agent.events.dropped.queue`: `COUNTER` of events where the attempt to add it to the reporting queue failed (e.g. due to a full queue)

>  - `agent.events.dropped.error`: `COUNTER` of events dropped after they previously have been successfully added to the queue.

Reviewer 2: Have you considered just having a `agent.events.dropped` metrics with a `reason` label?

Author: I used the metrics suggested in your space-time project as starting point ([document](https://docs.google.com/document/d/1Bi-Km-HRREWOtahgYyvVzQ5LzTD50Wfq-9feeTDExDU/edit#heading=h.7fsov1jn8ngv)).

I don't see a benefit / downside on either approach, so I'm free to change it if you have a preference.

Reviewer 2: If there's nothing that speaks against it, a label would be my preference. This also allows easier extension in the future, for example, differentiating different kinds of `error` scenarios, such as validation errors vs network errors.

All event count metrics MUST be able to be disabled by the `disable_metrics` configuration option.

### Event Queue Metrics

The following metrics SHOULD be recorded by agents regarding the state of the internal event reporting queue:

 - `agent.events.queue.min_size.pct`: The smallest queue usage in percent (range [0-1]) observed since this metric was reported the last time
 - `agent.events.queue.max_size.pct`: The biggest queue usage in percent (range [0-1]) observed since this metric was reported the last time

We capture both `min_size` and `max_size` to allow to distinguish between reasons for why the queue is becoming full:
 * There is too much data captured / network errors: Both `min_size` and `max_size` are at `100%`
 * The data is collected in bursts for which the queue is not correctly sized: `max_size` is at `100%`, but `min_size` is low.

The queue usage can be computed based on how the agent defines the queue capacity.
E.g. if the queue capacity is a fixed number of events, the usage can be computed based on the current number of events.
If the queue capacity is in bytes, the usage can be computed based on the number of bytes currently occupied in the queue.

> If agents use multiple queues (e.g. separating `metricsets` from other events), they SHOULD expose the `agent.events.queue.*` metrics for every queue and add a `queue_name` label.

Reviewer 2: Could it lead to issues if some agents don't define the `queue_name` label? I think it might but haven't thought it though. If so, we could require the `queue_name` label and if agents only have a single queue, they could set a name like `generic` or similar to indicate that this is a queue that holds multiple event types. This would also enable those agents to more easily add more queues in the future, without having to change the format of the metric.

The values for the `queue_name` label are implementation-defined.

All event queue metrics MUST be able to be disabled by the `disable_metrics` configuration option.

#### Possible implementation for agents

In order to track these metrics, most likely a custom implementation is required on top of the underlying queue.

One approach can be to have two variables `min_size` and `max_size`, which are atomically updated on changes to the queue:
 * When an item is added to the queue, update `max_size` with the current queue size if it is greater than `max_size`
 * When an item is removed from the queue, update `min_size` with the current queue size if it is less than `min_size`

>  * When an item cannot be added to the queue (dropped), set `max_size` to `queue_capacity`

Reviewer 2: ```suggestion
 * When an item cannot be added to the queue (dropped), set `max_size` to `1`
```

> After the metrics are exported, `min_size` and `max_size` need to be reset tu the current queue size.

Reviewer 2: ```suggestion
After the metrics are exported, `min_size` and `max_size` need to be reset to the current queue size.
```

## Agent Health Metrics

Agents SHOULD expose metrics which allow the analysis of the internal state of the agent, especially including the [reporting/transport mechanism](transport.md).

Many of these metrics are `COUNTER`s: a `COUNTER` is a monotonically increasing metric.
It is periodically reported by sending the difference of the counter-value now vs its value at the time of the previous report.

> In order to save elasticsearch disk space, zero-increases are not sent.

Reviewer 3: I know I'm late to this game, sorry @JonasKunz ;-)
Is the main reason for using counters here (instead of gauges) to potentially save disk space?
What if those events are dropped though? Doesn't this skew the information?

Author: The main reason is that it is currently difficult to visualize cumulative counter metrics in kibana. E.g. for showing the `agent.events.total` metric in Kibana, the query needs to first compute the increase of the counter for each `event_type` and afterwards sum them up.

There has been an offline discussion about this topic with @axw , where we felt like going with delta-based counters is fine for now. We did a very small brainstorming on possibilities for improving the reliability:
 * Buffering failed events and retry them on agent side
 * Switching to cumulative counters when TSDB is ready (by creating new metrics and deprecating the previous ones)
Both have pros and cons, we might even do a mixture of both.

Reviewer 3: I had a feeling that this was already discussed somewhere.
Thanks for the clarification!

Agents SHOULD attempt to assign the appropriate label value based on the counted event. If this would impose significant implementation overhead, the value `undefined` MUST be used instead.

The `agent.events.dropped` metric MUST have a value for the `reason` label. The following values MUST be used:
* `queue` MUST be used for events where the attempt to add it to the reporting queue failed (e.g. due to a full queue)
* `error` MUST be used for events dropped after they previously have been successfully added to the queue.

The value of `agent.events.dropped` with `reason=error` is an upper bound for the actual number of dropped events after entering the queue.
For example, when a request to the Intake API fails without a response, all events within this request MUST be considered as failed.
In reality, it is possible that e.g. half of the data was actually successfully received and forwarded by the APM server.
When the APM server responds with an HTTP error code, the number of dropped events SHOULD be computed by subtracting the `accepted` field from the response body from the total number of inflight events of this request.

> All event count metrics MUST be able to be disabled by the `disable_metrics` configuration option.

Reviewer 2: ```suggestion
It MUST be possible to disable all event count metrics via the `disable_metrics` configuration option.
```

### Event Queue Metrics

The following metrics SHOULD be recorded by agents regarding the state of the internal event reporting queue:

 - `agent.events.queue.min_size.pct`: The smallest queue usage in percent (range [0-1]) observed since this metric was reported the last time
 - `agent.events.queue.max_size.pct`: The biggest queue usage in percent (range [0-1]) observed since this metric was reported the last time

We capture both `min_size` and `max_size` to allow to distinguish between reasons for why the queue is becoming full:
 * There is too much data captured / network errors: Both `min_size` and `max_size` are at `100%`
 * The data is collected in bursts for which the queue is not correctly sized: `max_size` is at `100%`, but `min_size` is low.

The queue usage can be computed based on how the agent defines the queue capacity.
E.g. if the queue capacity is a fixed number of events, the usage can be computed based on the current number of events.
If the queue capacity is in bytes, the usage can be computed based on the number of bytes currently occupied in the queue.

> The `agent.events.queue.*` metrics must have a `queue_name` label. If agents use multiple queues, the `agent.events.queue.*` SHOULD be exposed for each queue with a implementation-defined value for `queue_name` per queue.

Reviewer 2: ```suggestion
The `agent.events.queue.*` metrics MUST have a `queue_name` label. If agents use multiple queues, the `agent.events.queue.*` SHOULD be exposed for each queue with a implementation-defined value for `queue_name` per queue.
```

If agents use just a single queue or have a shared primary queue, the value `generic` SHOULD be used as value for `queue_name` for this queue.

> All event queue metrics MUST be able to be disabled by the `disable_metrics` configuration option.

Reviewer 2: ```suggestion
It MUST be possible to disable all event queue metrics via the `disable_metrics` configuration option.
```

```java
AtomicLong maxSize = new AtomicLong();

public void updateMaxSize(long currentSize) {
    //alternatively abort after like 10 iterations for constant latency
    while (true) { 
        long current = maxSize.get();
        if (current >= currentSize) {
            return;
        }
        boolean casSuccess = maxSize.compareAndSet(current, currentSize);
        if (casSuccess) {
            return;
        }
    }
}

```

### Event Request Metrics

Agents SHOULD expose the following metrics regarding Intake API networking:

 - `agent.events.requests.count`: `COUNTER` of the number of requests made or attempted to the Intake API of the APM server
 - `agent.events.requests.bytes`: `COUNTER` of the approximate number of bytes on the wire sent to the Intake API of the APM

The `agent.events.requests.*` metrics MUST have the label `success`, which can have the values `true` or `false`. A request is counted with `success=true` iff the APM Server responded with `2xx`.

The metric `agent.events.requests.bytes` does not need to represent the exact network usage.
Instead, the number of compressed bytes within the request body can be used as approximation.

> All event request metrics MUST be able to be disabled by the `disable_metrics` configuration option.

Reviewer 2: ```suggestion
It MUST be possible to disable all event request metrics via the `disable_metrics` configuration option.
```

Events dropped before this attempt (e.g. due to [sampling](tracing-sampling.md) or [transaction_max_spans](handling-huge-traces/tracing-spans-limit.md)) are NOT counted.

The value of `agent.events.dropped.error` is an upper bound for the actual number of dropped events after entering the queue.
For example, when an request to the Intake API fails without a response, all events within this request MUST be considered as failed.
In reality, it is possible that e.g. half of the data was actually successfully received and forwarded by the APM server.
When the APM server responds with an HTTP error code, the number of dropped events SHOULD be computed by subtracting the `accepted` field from the response body from the total number of inflight events of this request.

The `agent.events.total` metric MUST have the label `eventType`. The possible values for `eventType` are the lowercase names of the events according to the [spec](https://github.com/elastic/apm-server/tree/main/docs/spec/v2) (e.g. `transaction`, `metricset`) or the value `other`.

> The `agent.events.dropped.*` metrics MUST NOT have any labels.

Author: I tried to design the metrics based on the questions they are supposed to answer when looking at them in a dashboard.

For the `agent.events.dropped.*` metrics I have the following questions in mind:
 * Did I not loose any data on the agent side? (sum of `agent.events.dropped.*` is zero)
 * How much data did I loose at maximum (pessimistic upper bound)? (sum of `agent.events.dropped.*`)
 * Was the data lost due to queue overload or other errors (network etc)? (`dropped.queue` vs `dropped.error`)
 
If both `agent.events.dropped.*` are labeled with the `eventType`, then it is possible to answer these questions for individual types of events. E.g. "I know I lost some data, but did I at least not loose any `metricsets`, so I can trust my metrics data? ".

If we now add the label `eventType` only to `dropped.queue` and not to `dropped.error`, I don't see any new valuable questions we could answer. The only one I can come up with is "How many transactions/metricsets did I loose due to the queue being full"? But I don't see any actual insight gained by this, e.g. If you didn't loose any metrics due to the queue being full it does not imply that you didn't loose any metrics at all and therefore you still can't really trust your metric data.

So, event though it is technically possible (and maybe easy) I still wouldn't capture the `eventType` for `dropped.queue`, because imo it doesn't provide more insights to the user viewing these metrics.
This is just my opinion, feel free to disagree here!

> The `agent.events.dropped.*` metrics MUST NOT have any labels.

Reviewer 4: I suppose let's just start without it, as you say; we can always add it later. Summing over all the types would be equivalent.

> The `agent.events.dropped.*` metrics MUST NOT have any labels.

Author: > One data point: in the Go Agent we have a separate queue/buffer for metrics.

Well that changes things up a bit for the queue utilization metrics. For which queue would you report `agent.events.queue.min_size.pct` and `agent.events.queue.max_size.pct`?
Options I see:
 * Add labels and expose the utilization per queue
 * Always use the non-metrics queue
   * Imagine an app which collects no traces but high cardinality metrics: metricsets might be dropped but the `max_size.pct` would remain at zero.
 * Compute MAX and MIN across the individual queues
   * Would imo render `min_size` useless, because the metrics queue very likely is almost always empty
 * Report the (min,max) utitlization of the queue, which has the highest max-usage in the current reporting time frame. Use a deterministic order if max-usage is equal across the queues
   * Imo the best solution here
   
Whichever solution we chose (might be none of the above), I think we should add it to the spec.

> The `agent.events.dropped.*` metrics MUST NOT have any labels.

Author: To elaborate a bit further after additional discussions at the offsite:

> Add labels and expose the utilization per queue

This is also not really feasible in my opinion.
First of all, I think that the metrics should reflect the knobs that the user is able to turn on the agents.
I think we have a single knob for tuning all of the queues at once in most agents (e.g. `ELASTIC_APM_API_BUFFER_SIZE` for the go agent, @axw please correct me if I'm wrong). Of course we as agent developers are interested in which specific queue is full. But we will use logs to get this information, not the metrics.

In addition, depending on the chosen labels, it might be hard to create a generic Kibana dashboard for all agent languages.
If we decide to add a label `queue-name`, where the label values are defined by each agent implementation, the only option (to my knowledge) in Kibana would be to display the utilization of all queues in a single chart via breakdown. However, this would be quiet confusing in comparison to having one char per queue I think.

Alternatively we could use `eventType` (span, transaction, etc) as label. This has significant downsides in case of queues shared for multiple event types, e.g. in the Java Agent.
For example, if the shared queue on the java agent is full, the `max_usage` metric would be on 100% for all `eventTypes`, even though it might be filled only with spans. When now looking at the `max_usage` for `metricsets`, this might lead users to the wrong conclusion that we need to reduce the number of metrics reported.

This is why I would propose this solution from my last comment as solution:
> Report the (min,max) utitlization of the queue, which has the highest max-usage in the current reporting time frame. Use a deterministic order if max-usage is equal across the queues

@axw @felixbarny Any thoughts / objections?

> The `agent.events.dropped.*` metrics MUST NOT have any labels.

Reviewer 2: > If we decide to add a label queue-name, where the label values are defined by each agent implementation, the only option (to my knowledge) in Kibana would be to display the utilization of all queues in a single chart via breakdown. 

I think we could visualize it by showing the max utilization of all queues which would be similar to 

> Report the (min,max) utitlization of the queue, which has the highest max-usage in the current reporting time frame. 

With the difference that we do this at the UI layer. This has the advantage that we can at least potentially drill in deeper (even if we don't have a dedicated UI for it in the first iteration).

I think the approach to report a utilization metric per queue gives us more optionality going forward with little impact on storage and complexity. But maybe I'm missing something.

> The `agent.events.dropped.*` metrics MUST NOT have any labels.

Author: >I think we could visualize it by showing the max utilization of all queues which would be similar to
>>Report the (min,max) utitlization of the queue, which has the highest max-usage in the current reporting time frame.

Regarding the `max_usage` this would work properly. However we would also get the `min_usage` across all queues instead of the queue which is currently under load. Therefore it is rather equal to this approach, but on visualization side:

> Compute MAX and MIN across the individual queues
Would imo render min_size useless, because the metrics queue very likely is almost always empty

I am okay with that as this would merely be a visualization limitation currently in kibana and we could always do a per-language dashboard with separate graphs per queue.

If you agree, I will add the optional `queueName` label to the spec with implementation-defined values per agent.

> The `agent.events.dropped.*` metrics MUST NOT have any labels.

Reviewer 4: Yes, snake_case should be preferred. I think some agents may have used camelCase, but it's not the norm.

> Agents SHOULD attempt to assign the appropriate label value based on the counted event. If this would impose significant implementation overhead, the value `other` MUST be used instead.

Author: Great suggestion! I also discovered that `other` slightly collides with the default `Other` label assigned by Kibana when breaking down metrics 😄

All event count metrics can be disabled via the `disable_metrics` configuration.

### Event Queue Metrics

The following metrics SHOULD be recorded by agents regarding the state of the internal event reporting queue:

 - `agent.events.queue.min_size.pct`: The smallest queue usage in percent (range [0-1]) observed since this metric was reported the last time

>  - `agent.events.queue.max_size.pct`: The biggest queue usage in percent (range [0-1]) observed since this metric was reported the last time

Author: > Do you have an idea in mind of how these metrics should be calculated? If we were to report just a single current utilisation metric it would be clear: just calculate it on the usual metrics interval.

Yes, this is how I've done it for the Java agent:
I have variables storing the current `minQueueSize` and `maxQueueSize`. Whenever the queue size changes (event is added or removed), these variables get updated with an [atomic min/max](https://github.com/JonasKunz/apm-agent-java/blob/93c3e6f765e3af41094a06bc71831b97d4877550/apm-agent-core/src/main/java/co/elastic/apm/agent/util/AtomicDouble.java#L52). In addition when an event gets dropped, I set `maxQueueSize` to `1.0`.

When the metrics are reported, the current values of `minQueueSize` and `maxQueueSize` are reported and the variables are reset to the current queue size.

> This seems a bit like [Pressure Stall Information](https://facebookmicrosites.github.io/psi/docs/overview). Here it would be more like the amount of time (over the reporting period) where the queue was full.

Yeah seems similar. However, if I understand it correctly, PSI aims to monitor induced latency due to a contended resource. We however do not induce any latency when reporting events to the queue, as we drop them instead of waiting due to a full queue. Monitoring the time the queue is full would be different to PSI: E.g. if your resource is full for 1 second, but no new task arrives, your PSI is still zero.

>  - `agent.events.queue.max_size.pct`: The biggest queue usage in percent (range [0-1]) observed since this metric was reported the last time

Reviewer 4: > Yes, this is how I've done it for the Java agent:
I have variables storing the current minQueueSize and maxQueueSize. Whenever the queue size changes (event is added or removed), these variables get updated with an [atomic min/max](https://github.com/JonasKunz/apm-agent-java/blob/93c3e6f765e3af41094a06bc71831b97d4877550/apm-agent-core/src/main/java/co/elastic/apm/agent/util/AtomicDouble.java#L52). In addition when an event gets dropped, I set maxQueueSize to 1.0.
>
> When the metrics are reported, the current values of minQueueSize and maxQueueSize are reported and the variables are reset to the current queue size.

Thanks, makes sense. Maybe that brief description to the spec, to assist other agent devs in their implementation?

> Monitoring the time the queue is full would be different to PSI: E.g. if your resource is full for 1 second, but no new task arrives, your PSI is still zero.

Good point!

>  - `agent.events.queue.max_size.pct`: The biggest queue usage in percent (range [0-1]) observed since this metric was reported the last time

Reviewer 2: > Whenever the queue size changes (event is added or removed), these variables get updated with an [atomic min/max](https://github.com/JonasKunz/apm-agent-java/blob/93c3e6f765e3af41094a06bc71831b97d4877550/apm-agent-core/src/main/java/co/elastic/apm/agent/util/AtomicDouble.java#L52).

Not necessarily a blocker but I'm a bit worried that this will induce latency overhead when a lot of threads are contending to add events to the queue. The disruptor queue we're using in the Java agent is carefully designed to minimize that overhead. Using a data structure that's designed around CAS loops to count min/max queue size could therefore add overhead when there's contention.
